server/expression/gms_cast.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	"github.com/dolthub/vitess/go/vt/proto/query"
+	"github.com/shopspring/decimal"
+
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// GMSCast handles the conversion from a GMS expression's type to its Doltgres type that is most similar.
+type GMSCast struct {
+	sqlChild sql.Expression
+}
+
+var _ sql.Expression = (*GMSCast)(nil)
+
+// NewGMSCast returns a new *GMSCast.
+func NewGMSCast(child sql.Expression) *GMSCast {
+	return &GMSCast{
+		sqlChild: child,
+	}
+}
+
+// Children implements the sql.Expression interface.
+func (c *GMSCast) Children() []sql.Expression {
+	return []sql.Expression{c.sqlChild}
+}
+
+// Child returns the child that is being cast.
+func (c *GMSCast) Child() sql.Expression {
+	return c.sqlChild
+}
+
+// DoltgresType returns the DoltgresType that the cast evaluates to. This is the same value that is returned by Type().
+func (c *GMSCast) DoltgresType() pgtypes.DoltgresType {
+	// GMSCast shouldn't receive a DoltgresType, but we shouldn't error if it happens
+	if t, ok := c.sqlChild.Type().(pgtypes.DoltgresType); ok {
+		return t
+	}
+	switch c.sqlChild.Type().Type() {
+	case query.Type_INT8, query.Type_INT16, query.Type_INT24, query.Type_INT32, query.Type_YEAR, query.Type_ENUM:
+		return pgtypes.Int32
+	case query.Type_INT64, query.Type_SET, query.Type_BIT, query.Type_UINT8, query.Type_UINT16, query.Type_UINT24, query.Type_UINT32:
+		return pgtypes.Int64
+	case query.Type_UINT64:
+		return pgtypes.Numeric
+	case query.Type_FLOAT32:
+		return pgtypes.Float32
+	case query.Type_FLOAT64:
+		return pgtypes.Float64
+	case query.Type_DECIMAL:
+		return pgtypes.Numeric
+	case query.Type_DATE, query.Type_DATETIME, query.Type_TIMESTAMP:
+		return pgtypes.Timestamp
+	case query.Type_TIME:
+		return pgtypes.Text
+	case query.Type_CHAR, query.Type_VARCHAR, query.Type_TEXT, query.Type_BINARY, query.Type_VARBINARY, query.Type_BLOB:
+		return pgtypes.Text
+	case query.Type_JSON:
+		return pgtypes.Json
+	case query.Type_NULL_TYPE:
+		return pgtypes.Null
+	case query.Type_GEOMETRY:
+		return pgtypes.Unknown
+	default:
+		return pgtypes.Unknown
```

Comment:
Should this be an error? Relying on other layers of the stack / analysis steps to catch?

Reply:
What does the error look like if I use an unsupported type? If the only user-facing error says "unsupported type: unknown", that's bad, and you should fix it.

Reply:
This is just for returning the type, since `Type()` doesn't return errors. Actual evaluation will error if we receive an unknown anywhere.